rfcs/20191017-tfx-standardized-inputs.md

```diff
@@ -600,68 +659,52 @@ be reused in a TF op (as the downstream of the Dataset, or in a Python wrapper).
-        computations (for example, slicing a StructuredTensor or computing the
-        mean) with StructuredTensors is through TF ops and their python
-        bindings.
-
-    Such a coupling is not in our favor because:
-
-    *   Some TFX components functionally do not require TF to work. For example,
-        TFDV can analyze any data set. TFMA can analyze a model while treating
-        the model as a blackbox. In both cases, the ML library that trained the
-        model is irrelevant and TF should not be assumed. And coupling with TF
-        does not only introduce a heavy dependency, but also forces the user to
-        learn about TF if they need to implement a TFXIO for their data format.
-
-    *   Operations against StructuredTensors boils down to Python + TF ops. The
-        overhead of either is much higher than just calling an Arrow C++ API
-        that does the same operation.
-
-    *   The main extension point of TF is Ops which don't understand the nested
-        structures. Compared to using Arrow’s C++ APIs, implementing an Op that
-        deals with StructuredTensors will be much complicated.
+## StructuredTensor
+
+We’ve considered an alternative where [StructuredTensor](https://github.com/tensorflow/community/blob/master/rfcs/20190910-struct-tensor.md)
+is the unified in-memory representation, but it does not meet all of the
+requirements:
+
+|                                   | Arrow RecordBatch  | StructuredTensor  |
+| :-------------------------------- | :----------------- | :---------------- |
+| Columnar                          | Yes                | Yes               |
+| Lossless encoding (nullity)       | Yes                | No (see remark 1) |
+| Efficient translatable to Tensors | Yes (see remark 2) | Yes               |
+| Efficient slicing, filtering and concatenation | Yes                | No (see remark 3) |
+| Interoperability with the rest of the world | Good through Apache Arrow | Needs adaptation (see remark 4) |
+
+Remarks:
+
+1.  We could revise the design of StructuredTensor to include the nullibility
+    support.
+2.  Only when the backing buffers are aligned correctly. Currently both TF
+    and Apache Arrow has 64-byte alignment. And this can be enforced by
+    implementing our own Arrow MemoryPool wrapping a TF allocator.
+3.  See the comparison in [this colab notebook](https://colab.sandbox.google.com/drive/1SCQs88J4Tc6HKk2AfpYvELaJDxa4h4IQ).
+    *   It’s worth calling out that Arrow is meant to be a data analysis library
+        and better data analysis support (for example, support for a “group-by”
+        clause) will be added over time.
+4.  We may gain similar interoperability by creating an Arrow to
+    StructuredTensor adapter.
+    *    Beyond the technical aspects, we believe by having all the TFX libraries
+         directly adopting a popular OSS in-memory format will send a positive
+         message that TFX is meant to work well with the rest of the world.
```

Comment:
I agree that this is the most compelling argument. Another thing to consider is hardware acceleration-- eg, via NVIDIA Rapids.